features/netsocket/Socket.h

```diff
@@ -21,93 +21,121 @@
-#include "netsocket/NetworkStack.h"
-#include "rtos/Mutex.h"
 #include "Callback.h"
-#include "mbed_toolchain.h"
-
 
 /** Abstract socket class
  */
 class Socket {
 public:
-    /** Destroy a socket
+    /** Destroy a socket.
      *
      *  Closes socket if the socket is still open
      */
     virtual ~Socket() {}
 
-    /** Opens a socket
+    /** Close the socket.
      *
-     *  Creates a network socket on the network stack of the given
-     *  network interface. Not needed if stack is passed to the
-     *  socket's constructor.
+     *  Closes any open connection and deallocates any memory associated
+     *  with the socket. Called from destructor if socket is not closed.
      *
-     *  @param stack    Network stack as target for socket
      *  @return         0 on success, negative error code on failure
      */
-    nsapi_error_t open(NetworkStack *stack);
+    virtual nsapi_error_t close() = 0;
 
-    template <typename S>
-    nsapi_error_t open(S *stack) {
-        return open(nsapi_create_stack(stack));
-    }
-    
-    /** Close the socket
+    /** Connects socket to a remote host.
      *
-     *  Closes any open connection and deallocates any memory associated
-     *  with the socket. Called from destructor if socket is not closed.
+     *  Initiates a connection to a remote server specified by the
+     *  indicated address. In case of connectionless protocol, set
+     *  the remote address for next send() call.
      *
+     *  @param address  The SocketAddress of the remote host
      *  @return         0 on success, negative error code on failure
      */
-    nsapi_error_t close();
-    
-    /** Subscribes to an IP multicast group
+    virtual nsapi_error_t connect(const SocketAddress &address) = 0;
+
+    /** Send data on a socket
+     *
+     *  The socket must be connected to a remote host before send() call.
+     *  Returns the number of bytes sent from the buffer.
+     *  In case of connectionless socket, send data to pre-specified remote.
      *
-     * @param address   Multicast group IP address
-     * @return          Negative error code on failure
+     *  By default, send blocks until all data is sent. If socket is set to
+     *  non-blocking or times out, a partial amount can be written.
+     *  NSAPI_ERROR_WOULD_BLOCK is returned if no data was written.
+     *
+     *  @param data     Buffer of data to send to the host
+     *  @param size     Size of the buffer in bytes
+     *  @return         Number of sent bytes on success, negative error
+     *                  code on failure.
      */
-    int join_multicast_group(const SocketAddress &address);
+    virtual nsapi_size_or_error_t send(const void *data, nsapi_size_t size) = 0;
 
-    /** Leave an IP multicast group
+    /** Receive data from a socket.
+     *
+     *  Receive data from connected socket or in case of connectionless socket
+     *  this is equivalent of calling recvfrom(NULL, data, size).
```

Comment:
I still can't decide whether we should actually have final implementations of `send`/`recv` here to enforce the equivalence.

Reply:
Do you mean that `send()` would be just `sendto(NULL, data, size);`

Reply:
That's what I'm pondering, but looking more closely at the POSIX specs I actually don't see the `sendto(NULL)` == `send()` equivalence - only `send(0)` == `write` and connection-mode `sendto()` == `send()`.